drivers/flash/flash_cadence_nand_ll.c

```diff
@@ -0,0 +1,1516 @@
+{
+
+	uint32_t mini_ctrl_cmd_l, mini_ctrl_cmd_h, status;
+	uintptr_t base_address;
+	int ret = 0;
+
+	base_address = params->nand_base;
+	mini_ctrl_cmd_l = mini_ctrl_cmd & U32_MASK_VAL;
+	mini_ctrl_cmd_h = mini_ctrl_cmd >> 32;
+	ret = cdns_nand_wait_idle(base_address);
+
+	if (ret != 0) {
+		LOG_ERR("Wait for controller to be in idle state Failed");
+		return ret;
+	}
+	sys_write32(mini_ctrl_cmd_l, (base_address + CDNS_CMD_REG2));
+	sys_write32(mini_ctrl_cmd_h, (base_address + CDNS_CMD_REG3));
+	/* Select generic command. */
+	status = CMD_0_THREAD_POS_SET(NF_TDEF_TRD_NUM);
+#ifdef CONFIG_CDNS_NAND_INTERRUPT_SUPPORT
+	status |= GEN_CF_INT_SET(GEN_CF_INT_ENABLE);
+#endif
+	status |= CMD_0_C_MODE_SET(CT_GENERIC_MODE);
+	sys_write32(status, (base_address + CDNS_CMD_REG0));
+	return 0;
+}
+
+/**
+ * Send a generic command data to the Cadence NAND controller.
+ *
+ * @param params The Cadence NAND parameters structure.
+ * @param mode The mode of operation (read, write).
+ * @param data_length The length of the associated data.
+ * @retval   0 on success or -ETIMEDOUT error value on failure.
+ */
+static int cdns_generic_cmd_data(struct cadence_nand_params *params, uint8_t mode,
+				 uint32_t data_length)
+{
+	uint64_t mini_ctrl_cmd = 0;
+	int ret = 0;
+
+	mini_ctrl_cmd |= GCMD_TWB_VALUE;
+	mini_ctrl_cmd |= GCMCD_DATA_SEQ;
+	mini_ctrl_cmd |= GEN_SECTOR_COUNT_SET;
+	mini_ctrl_cmd |= GEN_LAST_SECTOR_SIZE_SET((uint64_t)data_length);
+	mini_ctrl_cmd |= GEN_DIR_SET((uint64_t)mode);
+	mini_ctrl_cmd |= GEN_SECTOR_SET((uint64_t)data_length);
+	ret = cdns_generic_send_cmd(params, mini_ctrl_cmd);
+	return ret;
+}
+
+/**
+ * Wait for the completion of an SDMA operation in the Cadence NAND controller.
+ *
+ * @param base_address The base address of the Cadence NAND controller.
+ * @retval  0 on success or -ETIMEDOUT error value on failure.
+ */
+static int cdns_wait_sdma(uintptr_t base_address)
+{
+
+	if (!WAIT_FOR(((sys_read32(base_address + INTR_STATUS) & BIT(SDMA_TRIGG)) != 0),
+		      IDLE_TIME_OUT, k_msleep(1))) {
+		LOG_ERR("Timed out waiting for sdma response");
+		return -ETIMEDOUT;
+	}
+	sys_set_bit((base_address + INTR_STATUS), SDMA_TRIGG);
+	return 0;
+}
+
+/**
+ * Perform buffer copying from/to SDMA regs in the Cadence NAND controller.
+ *
+ * @param sdma_base_address The base address of the SDMA in the Cadence NAND controller.
+ * @param buffer The source or destination buffer for the copy operation.
+ * @param data_length The length of the data to be copied.
+ * @param dir The direction of the copy operation (read or write).
+ */
+static void sdma_buffer_copy(uint32_t sdma_base_address, uint8_t *buffer, uint32_t data_length,
+			     uint8_t dir)
+{
+	uint32_t index;
+
+	if (dir == CDNS_READ) {
+		for (index = 0; index < data_length; index++) {
+			*(buffer + index) = sys_read8(sdma_base_address + index);
+		}
+	} else {
+		for (index = 0; index < data_length; index++) {
+			sys_write8(*(buffer + index), sdma_base_address + index);
+		}
+	}
+}
```

Comment:
Not insisting but you could probably have just copyin/copyout fuctions for that.
I doubt we save any flash by stacking both directions here, assuming that the dir has to be passed to a function which means that each call has to pass it through some register, so probably save contents of the register on stack, and so on.

Reply:
copyin and copyout two seprate function added.

Reply:
@de-nordic  i fixed all review comments could you please review once?